### ISSUE TITLE
Off by one energy bar (D2X)

### ISSUE THREAD

**Issue by zicodxx** [dxx-rebirth/dxx-rebirth#169]
This seems to be related to 90e3c294643111692b7c6873fa68dc973e827ac8. This commit fixes the off-by-one for D1X (PC and MAC version) however breaks the D2X status bar in 320x200 as well as the 640x480 hires graphics - which now has a constant yellow line on the left while the bar empties.

This is kinda odd since the commit looks as if it should work - both the graphics and the box drawing over it should be at the same coordinates. It's possible that I added some +/-1 offsets for certain functions back in the day to make the HUD scale more nicely to all possible resolutions - despite texture filtering and rounding. And it's also possible I left that out for D1X as I didn't have the Hires Mac for it back in the day and this is now causing this exact issue. However I cannot find the origin of it right now which is why I'm posting it here.

**Comment by ryusei117** [comment]
Strange, I couldn't reproduce this new bug ... I tried 320 x 200, 640 x 480, and 640 x 480 with "-lowresgraphics" on the command line. Is there a specific setting that triggers the bug - such as multisampling? I wonder if this could be system-specific ... I was testing on Debian GNU/Linux 8.2 amd64 on my laptop with an Intel integrated graphics processor.

**Comment by zicodxx** [comment]
I uploaded an example shot right there: http://dxx-rebirth.com/tmp/d2x-offbyone.png
This has multisampling enabled but it's the same with it being disabled. I know texture filtering can cause this depending on the resolution scale which is why I sometimes added these +/- .5 to 1 offsets. I assume the problems stems from gr_uline in this case. Examining the 640x480 screenshot, the graphic starts on pixel 196 (on x) while the deleting line starts at 197 while SB_ENERGY_GAUGE_X is defined with 196 for 640x480. So it's either a difference in i2f or gr_uline in D2X as compared to D1X.

I have not investigated this much further as I have two little other tasks on my mind and I do not want to take this lightly as I think fixing this needs a lot of testing. Depending on if it's gr_uline or i2f, this may make adjustments to many other HUD elements necessary (or it fixes more than it breaks. Again, needs to be tested).

**Comment by zicodxx** [comment]
I have to declare this one as invalid. I don't know where I screwed up but I must've to. By all means, even going back to the revision of that day I am unable to reproduce this bug. Using the same settings as shown in the screenshot, the energy gauge looks correct now.

**Comment by zicodxx** [comment]
While testing for this (https://github.com/CDarrow/DXX-Retro/issues/65), I started seeing the bug I initially reported here, which is why I'm going to re-open this. However while my initial observation was on D2X with an unscaled Status Bar (640x480), now I see it on a scaled (also 640x480) Status Bar in D1X. As @ryusei117 pointed out, it's related to Multisampling as now it does properly go away with the feature being disabled. In my initial observation in D2X it did not go away by disabling Multisampling (unless I made a mistake disabling it and instead toggled another feature on accident). 
Right now, I can see this with todays revision 5b9ccda8e85864367fcf5ecc3330301aaa12a20c, D1X running at 640x480 with Multisampling enabled. Will track this down and see where this comes from.

**Comment by zicodxx** [comment]
I have committed 081e5d5e3cf75905b2ef545b54726fcafdab9eeb which *should* fix the issue in most cases. What I did was to remove the rounding offsets in ogl_ulinec for X variables. The 4x Multisampling in the game seems to interpolate stronger on X than on Y and to the right (as compared via screencaps). Naturally this doesn't mean it's perfect on all resolutions or all possible FSAA methods. But without any filtering enabled, it seems to be accurate on all resolutions from 320x200 to 1920x1080 that my monitor supports.

I'll keep this issue open for a few more days in case someone sees some negative side effects.
